MdeModulePkg/Core/Dxe/Misc/MemoryProtection.c

```diff
@@ -529,7 +632,7 @@ ProtectUefiImage (
 
   if ((ProtectionPolicy == PROTECT_IF_ALIGNED_ELSE_ALLOW)) {
     ClearAccessAttributesFromMemoryRange (
-      (EFI_PHYSICAL_ADDRESS)(UINTN)LoadedImage->ImageBase,
+      (EFI_PHYSICAL_ADDRESS)(UINTN)LoadedImage->ImageBase & ~EFI_PAGE_MASK,

```

Comment:
If the ImageBase is unaligned, this might end up clearing something else that was also included in the same page.

i..e ImageBase = 0xBBB953C0 it would clear the attributes starting from BB95000 - BB95FFFF. But If something was in the memory before that, chances are it was unprotected as well.

Reply:
Correct, this is intentional. If unaligned images are used, we have to clear XP from the entire page(s). This will affect whatever else is there, but only in a freeing sense, not restrictive. So either another unaligned image, in which case it also needs it or random memory, in which case marking it RWX won’t cause it to crash, just degrade the security state. But, the system already has a degraded security state because of this unaligned image and it will crash if NX is not removed.

The general advice would be to align all images to either default or runtime page granularity, depending on driver type.